src/backend_model.cc

```diff
@@ -268,25 +266,26 @@ TritonModel::UpdateInstanceGroup(
       &model_config));
   RETURN_IF_ERROR(ValidateInstanceGroup(model_config, min_compute_capability_));
 
-  // Update the instances to the new config.
-  caller_lock->unlock();  // allow inference while creating instances
-  Status status = TritonModelInstance::SetInstances(
-      this, backend_cmdline_config_map_, host_policy_map_, model_config);
-  caller_lock->lock();
+  // Prepare the new instances on the new config.
+  std::vector<std::shared_ptr<TritonModelInstance>> added_instances,
+      removed_instances;
+  Status status =
+      PrepareInstances(model_config, &added_instances, &removed_instances);
+  if (!status.IsOk()) {
+    ClearBackgroundInstances();
+    return status;
+  }
+
+  // Update the scheduler.
+  status = UpdateConfiguredScheduler(added_instances, removed_instances);
   if (!status.IsOk()) {
-    // Remove any pending instances if created.
-    bg_instances_.clear();
-    bg_passive_instances_.clear();
+    ClearBackgroundInstances();
     return status;
   }
 
-  // At this point, the new model config is ready but not yet written into this
-  // object. The 'caller_lock' is held, so 'model_lifecycle' will pause any new
-  // inference request. It is safe to move forward and commit the change.
-  config_.mutable_instance_group()->Swap(model_config.mutable_instance_group());
-  RETURN_IF_ERROR(CommitInstances());
-  // Only model owned dynamic batch scheduler can be updated currently, so there
-  // is no need to update the scheduler.
+  // Commit the instance update.
+  CommitInstances();
+  *config_.mutable_instance_group() = model_config.instance_group();
```

Comment:
I noticed the order of these two operations switched, and that you went from `->Swap()` to `=`. Any implications?

Reply:
> the order of these two operations switched

The `CommitInstances();` use to be `RETURN_IF_ERROR(CommitInstances());`. If an error is returned and rolled-back during commit, the model config should not be changed. However, `CommitInstances();` should not (and does not) return any error, so the `RETURN_IF_ERROR()` is removed and the order is kept as it is.

I do not think swapping those two lines will make any difference, because updating the instance group is for the next round of model instance update, which cannot happen concurrently with this update, as it is safeguarded by model repository manager.

> went from `->Swap()` to `=`

I saw a segfault in the past that might be attributed to updating the model config protobuf message (swapping while accessing part of the message). From my readings, protobuf is not thread safe unless otherwise noted, and the only note I found is only [allocation methods are thread safe](https://protobuf.dev/reference/cpp/arenas/#thread-safety). Swapping is not an allocation, but assignment involves deep copy which could be regarded as an allocation. This is still more on the experiment side to see if the issue is fixed, but I have not noticed any related failures on this scheduler update branch after switching from swap to assignment.

Reply:
> I saw a segfault in the past that might be attributed to updating the model config protobuf message (swapping while accessing part of the message).

Shall we comment a note that this line may be cause for concern?

Reply:
The segfault could relate to something else than the protobuf message swapping. I think leaving a note might lead debugging into the wrong way, if that would happen again.